[New] `checkContextObjects` option in `display-name`

JulesBlm opened the pull request:
I'm proposing adding a new rule that checks if the `displayName` property is set for context objects, see [`Context.displayName`](https://reactjs.org/docs/context.html#contextdisplayname) in the React docs.

With multiple Context providers in an app, consistently setting a `displayName` makes it easy to see which is which in the React dev tools.

Some things to consider

* Should it perhaps be part of the `display-name` rule instead of a separate rule?
* Does it need more tests?
* Does this checking approach make sense? (This is my first time writing an ESLint rule, so I wouldn't be surprised if this is not the best way)
* Does it need to check more cases?

ljharb commented on tests/lib/rules/context-display-name.js:
```diff
@@ -0,0 +1,111 @@
+
+        const Hello = React.createContext();
+
+        const obj = {};
+        obj.displayName = "False positive";
```
should there be some of these test cases that avoid React.createContext entirely?

JulesBlm replied:
Just like this?

```javascript
const obj = {};
obj.displayName = "False positive";
```

ljharb replied:
yup, looks good!

JulesBlm replied:
OK added!

ljharb reviewed (commented):
Overall, this seems reasonable. Should this maybe be an option on the `display-name` rule?

JulesBlm commented:
> Overall, this seems reasonable. Should this maybe be an option on the display-name rule?

I think that makes sense, the intention is the same. Main difference is that `createContext` does not create a component directly but a context object containing two components.
I looked at the implementation of `display-name` and it makes heavy use of the `Components` utility, which can't be used for `createContext`. I thought it would be easier and cleaner to make a separate rule for it but I'm not married to the idea of it being a separate rule. What would your preference be?

ljharb commented:
while that's a fair point, since the rule is "display name" and not "component display name" i think it's worth figuring out the complexity inside the rule, rather than adding complexity for users.

JulesBlm commented:
> while that's a fair point, since the rule is "display name" and not "component display name" i think it's worth figuring out the complexity inside the rule, rather than adding complexity for users.

Alright makes sense, I turned it into a `checkContextObjects` option in `display-name`. As far as I can tell, the logic for detecting component isn't of much use for detecting context definitions so I chose not make use of the `Components` utility for it.

ljharb reviewed (commented):
There's a few lines missing test coverage; otherwise it looks pretty great!

JulesBlm commented on lib/rules/display-name.js:
```diff
@@ -87,6 +124,18 @@ module.exports = {
+     * @param {Object} contextObj The context object to process
+     */
+    function reportMissingContextDisplayName(contextObj) {
+      if (testReactVersion(context, '>= 16.3.0')) {
```
Yes sure, I moved it in `checkContextObjects` like this:

```js
const checkContextObjects = (config.checkContextObjects || false) && testReactVersion(context, '>= 16.3.0');
```

JulesBlm commented:
> There's a few lines missing test coverage; otherwise it looks pretty great!

Thanks! I noticed I forgot to enable the new flag for the new `valid` tests so I added that. I added few more tests too, some formatting fixes and moved the React version check.

I'm a little confused how to increase the test coverage. What would need to be done to meet the threshold? Test the `isCreateContext` util?

ljharb commented on lib/rules/display-name.js:
```diff
@@ -70,6 +77,36 @@ module.exports = {
+     */
+    function isCreateContext(node) {
+      if (!node.init) {
+        return false;
```
this line is what we'd need to hit - when would `node` have a falsy `init` value?

JulesBlm replied:
Initially I put it in here as a guard without much thought. This comment got me thinking that contexts declared like this were not covered by the rule

```

JulesBlm replied:
Initially I put this line in as guard without much thought. Thinking about it, `node.init` can only be falsy with a variable declaration without an initial value eg `let Hello;`. This got me thinking about the following: The rule would not detect contexts in assignments like this

```js
import { createContext } from 'react';

var Hello = {};
Hello = React.createContext(); // or createContext()
```

While unusual, I added a check for this too along with tests for it.

ljharb replied:
Awesome, thanks!

ljharb commented on lib/rules/display-name.js:
```diff
@@ -70,6 +77,58 @@ module.exports = {
+     * @param {ASTNode} node - The AST node being checked.
+     * @returns {Boolean} - True if node is a React.createContext call, false if not.
+     */
+    function isCreateContext(node) {
```
this seems like a good idea to add to `Components.js`, so other rules could share it?

ljharb replied:
Sure, a separate file is fine too, it just may need to be in the things provided to `Components.detect`

ljharb reviewed (commented):
Looks good overall

JulesBlm commented:
Hey, is there anything I can do to move this PR forward?

ljharb commented:
@JulesBlm sorry for the delayed reply; i replied to your comment above and will review today.

ljharb reviewed (approved):
This looks great - I'll land this later tonight.

JulesBlm commented:
Awesome thanks for your review @ljharb!

christianvuerings commented:
@ljharb https://www.npmjs.com/package/eslint-plugin-react still shows `7.32.2` as the latest version. Did anything go wrong during the npm publish process?
![image](https://github.com/jsx-eslint/eslint-plugin-react/assets/127199/2625229b-0ea6-42a2-bdbf-2d879d7b9420)

ljharb commented:
huh, maybe so, let me check

ljharb commented:
ah thanks, published it manualy

christianvuerings commented:
@ljharb that worked, thank you!
